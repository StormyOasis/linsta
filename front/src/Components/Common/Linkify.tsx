import React from 'react';
import parse, {
    DOMNode,
    domToReact,
    Element,
    Text as HtmlTextNode,
} from 'html-react-parser';
import StyledLink from './StyledLink';


const parseStyleString = (styleString: string): React.CSSProperties => {
    return styleString.split(';').reduce((styleObj, declaration) => {
        const [property, value] = declaration.split(':').map(s => s.trim());
        if (property && value) {
            const camelCaseProp = property.replace(/-([a-z])/g, (_, char) => char.toUpperCase());
            (styleObj as any)[camelCaseProp] = value;
        }
        return styleObj;
    }, {} as React.CSSProperties);
}

type LinkifyProps = {
    html: string;
    onClick?: (e:React.MouseEvent) => void;
}

const Linkify: React.FC<LinkifyProps> = (props: LinkifyProps) => {
    const linkifyText = (text: string): React.ReactElement[] => {
        const regex = /(@[a-zA-Z0-9._]+|#[a-zA-Z0-9_]+)/g;
        const parts = text.split(regex);

        return parts.map((part, index) => {
            if (part.startsWith('@')) {
                const username = part.substring(1);
                return (
                    <StyledLink key={index} to={`/${username}`} onClick={props.onClick}>
                        {part}
                    </StyledLink>
                );
            } else if (part.startsWith('#')) {
                return (
                    <StyledLink key={index} to={`/explore?q=${encodeURIComponent(part)}`} onClick={props.onClick}>
                        {part}
                    </StyledLink>
                );
            } else {
                return <React.Fragment key={index}>{part}</React.Fragment>;
            }
        });
    };

    const transform = (
        node: DOMNode,
        index: number
    ): string | boolean | void | React.ReactElement | null => {
        if (node.type === 'text') {
            const textNode = node as HtmlTextNode;
            return <React.Fragment key={index}>{linkifyText(textNode.data)}</React.Fragment>;
        }

        if (node.type === 'tag') {
            const element = node as Element;
            const { attribs, name, children } = element;

            const props: { [key: string]: any } = { ...attribs };

            // Convert class to className
            if ('class' in props) {
                props['className'] = props['class'];
                delete props['class'];
            }

            // Convert style string to React style object
            if ('style' in props) {
                props['style'] = parseStyleString(props['style']);
            }

            return React.createElement(
                name,
                { ...props, key: index },
                domToReact(children as DOMNode[], { replace: transform })
            );
        }

        return null;
    };

    return <>{parse(props.html, { replace: transform })}</>;
};

export default Linkify;
