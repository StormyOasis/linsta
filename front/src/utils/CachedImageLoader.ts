import { clear, del, get, set } from 'idb-keyval';
import { blobToBase64 } from "./utils";

export const loadImageCached = async (id: string, url: string): Promise<string> => {
    // First try to get the image from the cache
    const value = await get(id);

    if(value === undefined) {
        // Image not in the cache so have to fetch it
        try {
            const result = await fetch(url);
            if(result.status === 200) {
                const blob = await result.blob();
                const b64:string = await blobToBase64(blob) as string;

                // add the base64 data into the cache and return it
                await set(id, b64);

                return b64;
            }
        } catch(err) {
            console.log(err);
            return url; //default to what was passed in
        }

    } else {
        // Image is in the cache already as a base64 string, use that as is
        return value;
    }

    return url; //default to what was passed in
}

export const setImage = async (id: string, data: string) => {     
    await set(id, data);
}

export const deleteFromCache = async (id: string) => {
    // First try to get the image from the cache
    const value = await get(id);

    if(value === undefined) {
        // key not present in cache so do nothing
        return;
    }

    // delete from the cache
    await del(id);
}

export const clearCache = async () => {
    await clear();
}