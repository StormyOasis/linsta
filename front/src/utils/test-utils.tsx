import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { RootState } from "../Components/Redux/redux";
import { authInitialState } from '../Components/Redux/slices/auth.slice';
import { modalInitialState } from '../Components/Redux/slices/modals.slice';
import { profileInitialState } from '../Components/Redux/slices/profile.slice';
import { miscInitialState } from '../Components/Redux/slices/misc.slice';
import { configureStore } from '@reduxjs/toolkit';

import Theme from '../Components/Themes/Theme';

export const mockInitialState = (overrides?: Partial<RootState>): RootState => ({
    auth: { ...authInitialState, ...(overrides?.auth ?? {}) },
    modal: { ...modalInitialState, ...(overrides?.modal ?? {}) },
    profile: { ...profileInitialState, ...(overrides?.profile ?? {}) },
    misc: { ...miscInitialState, ...(overrides?.misc ?? {}) },
});

type ExtendedRenderOptions = {
    store?: ReturnType<typeof configureStore>;
    preloadedState?: any;
    route?: string; // ‚Üê allow overriding the initial route
} & Omit<RenderOptions, "queries">;

const dummyAuthReducer = (state = { user: null, isLoading: false, error: null }, _action: any) => state;
const dummyModalReducer = (state = { isOpen: false }, _action: any) => state;
const dummyProfileReducer = (state = { name: '', email: '' }, _action: any) => state;
const dummyMiscReducer = (state = { someData: null }, _action: any) => state;

const renderWithStore = (
    ui: ReactElement,
    {
        preloadedState,
        store = configureStore({
            reducer: {
                auth: dummyAuthReducer,
                modal: dummyModalReducer,
                profile: dummyProfileReducer,
                misc: dummyMiscReducer,
            },
            preloadedState,
        }),
        route = "/",
        ...renderOptions
    }: ExtendedRenderOptions = {}
) => {
    const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
        return (
            <Provider store={store}>
                <MemoryRouter initialEntries={[route]}>
                    <Theme>{children}</Theme>
                </MemoryRouter>
            </Provider>
        );
    };

    return {
        ...render(ui, { wrapper: Wrapper, ...renderOptions }),
        store,
    };
};

export * from "@testing-library/react";
export { renderWithStore };
export type { ExtendedRenderOptions };