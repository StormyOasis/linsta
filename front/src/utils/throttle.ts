import { useRef } from 'react';

const useThrottle = (callback: Function, delay: number) => {
    const lastCall = useRef(0);

    return (...args: any[]) => {
        const now = new Date().getTime();
        if (now - lastCall.current >= delay) {
            lastCall.current = now;
            callback(...args);
        }
    };
};

export default useThrottle;