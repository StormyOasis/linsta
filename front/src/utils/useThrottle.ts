import { useRef, useCallback } from 'react';

/**
 * React hook for throttling a callback function.
 * The callback will only be invoked at most once every `delay` milliseconds.
 *
 * @param callback - The function to throttle.
 * @param delay - The minimum delay (ms) between calls.
 * @returns A throttled version of the callback.
 */
const useThrottle = <T extends (...args: any[]) => void>(callback: T, delay: number) => {
    const lastCall = useRef<number>(0);

    return useCallback(
        (...args: Parameters<T>) => {
            const now = performance.now();
            if (now - lastCall.current >= delay) {
                lastCall.current = now;
                callback(...args);
            }
        },
        [callback, delay]
    );
};

export default useThrottle;