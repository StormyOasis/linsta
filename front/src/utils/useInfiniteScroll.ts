import React, { useEffect, useRef } from 'react';
import useThrottle from './useThrottle';

/**
 * Type for the async load more function.
 */
type AsyncLoadMoreFunction = () => Promise<void>;

/**
 * Custom hook for infinite scrolling.
 * Attaches a throttled scroll event listener to a container and triggers `loadMore`
 * when the user scrolls near the bottom.
 *
 * @param loadMore - Async function to load more items.
 * @param isLoading - Whether a load is currently in progress.
 * @param containerRef - Ref to the scrollable container element.
 */
const useInfiniteScroll = (loadMore: AsyncLoadMoreFunction, 
    isLoading: boolean, 
    childRef: React.MutableRefObject<HTMLDivElement | null>) => {
        
    const hasScrolled = useRef<boolean>(false);  // Track if the user has scrolled

    // Throttled scroll handler
    const throttledHandleScroll = useThrottle(async () => {
        if (typeof window !== 'undefined' && childRef.current) {
            const element = childRef.current as HTMLElement;
            const currentScroll = window.innerHeight + element.scrollTop;

            if (currentScroll + 256 >= element.scrollHeight && !isLoading && !hasScrolled.current) {
                hasScrolled.current = true;  // Mark as scrolled at least once
                await loadMore();
            }
        }
    }, 200);

    useEffect(() => {
        const childElement = childRef.current;
        if (childElement) {
            childElement.addEventListener('scroll', throttledHandleScroll);
        }

        // Clean up event listener on unmount
        return () => {
            if (childElement) {
                childElement.removeEventListener('scroll', throttledHandleScroll);
            }
        };
    }, [throttledHandleScroll, childRef]);

    // Reset trigger flag when loading finishes
    useEffect(() => {
        if (!isLoading) {
            hasScrolled.current = false;
        }
    }, [isLoading]);
}

export default useInfiniteScroll;