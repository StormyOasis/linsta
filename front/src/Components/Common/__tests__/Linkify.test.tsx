import React from "react";
import { screen, fireEvent } from "@testing-library/react";
import Linkify from "../Linkify";
import '@testing-library/jest-dom';
import { renderWithStore } from '../../../utils/test-utils';

// Mock StyledLink to a simple anchor for easier testing
jest.mock("../StyledLink", () => (props: any) => (
    <a data-testid="styled-link" href={props.to} onClick={props.onClick}>{props.children}</a>
));

describe("Linkify", () => {
    it("renders plain text without links", () => {
        const r = renderWithStore(<Linkify html="Hello world!" />);
        expect(screen.getByText("Hello world!")).toBeInTheDocument();
        expect(screen.queryByTestId("styled-link")).not.toBeInTheDocument();
        expect(r).toMatchSnapshot();
    });

    it("linkifies @mentions", () => {
        const r = renderWithStore(<Linkify html="Hello @alice!" />);
        const link = screen.getByTestId("styled-link");
        expect(link).toHaveAttribute("href", "/alice");
        expect(link).toHaveTextContent("@alice");
        expect(r).toMatchSnapshot();
    });

    it("linkifies #hashtags", () => {
        const r = renderWithStore(<Linkify html="Check #reactjs" />);
        const link = screen.getByTestId("styled-link");
        expect(link).toHaveAttribute("href", "/explore?q=%23reactjs");
        expect(link).toHaveTextContent("#reactjs");
        expect(r).toMatchSnapshot();
    });

    it("renders multiple links and plain text", () => {
        const r = renderWithStore(<Linkify html="Hi @bob, check #js and @carol!" />);
        const links = screen.getAllByTestId("styled-link");
        expect(links[0]).toHaveAttribute("href", "/bob");
        expect(links[0]).toHaveTextContent("@bob");
        expect(links[1]).toHaveAttribute("href", "/explore?q=%23js");
        expect(links[1]).toHaveTextContent("#js");
        expect(links[2]).toHaveAttribute("href", "/carol");
        expect(links[2]).toHaveTextContent("@carol");
        expect(r).toMatchSnapshot();
    });

    it("calls onClick when a link is clicked", () => {
        const handleClick = jest.fn();
        const r = renderWithStore(<Linkify html="Hello @alice!" onClick={handleClick} />);
        const link = screen.getByTestId("styled-link");
        fireEvent.click(link);
        expect(handleClick).toHaveBeenCalled();
        expect(r).toMatchSnapshot();
    });

    it("parses and applies inline styles and classes", () => {
        const r = renderWithStore(<Linkify html='<span class="my-class" style="color: red; font-weight: bold;">Styled @bob</span>' />);
        const span = screen.getByText(/Styled/).closest("span");
        expect(span).toHaveClass("my-class");
        expect(span).toHaveStyle({ color: "red", fontWeight: "bold" });
        // The @bob mention inside should be linkified
        const link = screen.getByTestId("styled-link");
        expect(link).toHaveAttribute("href", "/bob");
        expect(link).toHaveTextContent("@bob");
        expect(r).toMatchSnapshot();
    });

    it("parses nested HTML and linkifies text nodes", () => {
        const r = renderWithStore(<Linkify html='<div>Hello <b>@alice</b> and <i>#react</i></div>' />);
        const links = screen.getAllByTestId("styled-link");
        expect(links[0]).toHaveAttribute("href", "/alice");
        expect(links[0]).toHaveTextContent("@alice");
        expect(links[1]).toHaveAttribute("href", "/explore?q=%23react");
        expect(links[1]).toHaveTextContent("#react");
        expect(r).toMatchSnapshot();
    });
});