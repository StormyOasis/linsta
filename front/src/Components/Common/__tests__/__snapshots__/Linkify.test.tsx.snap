// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Linkify calls onClick when a link is clicked 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      Hello 
      <a
        data-testid="styled-link"
        href="/alice"
      >
        @alice
      </a>
      !
    </div>
  </body>,
  "container": <div>
    Hello 
    <a
      data-testid="styled-link"
      href="/alice"
    >
      @alice
    </a>
    !
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`Linkify linkifies #hashtags 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      Check 
      <a
        data-testid="styled-link"
        href="/explore?q=%23reactjs"
      >
        #reactjs
      </a>
    </div>
  </body>,
  "container": <div>
    Check 
    <a
      data-testid="styled-link"
      href="/explore?q=%23reactjs"
    >
      #reactjs
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`Linkify linkifies @mentions 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      Hello 
      <a
        data-testid="styled-link"
        href="/alice"
      >
        @alice
      </a>
      !
    </div>
  </body>,
  "container": <div>
    Hello 
    <a
      data-testid="styled-link"
      href="/alice"
    >
      @alice
    </a>
    !
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`Linkify parses and applies inline styles and classes 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="my-class"
        style="color: red; font-weight: bold;"
      >
        Styled 
        <a
          data-testid="styled-link"
          href="/bob"
        >
          @bob
        </a>
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="my-class"
      style="color: red; font-weight: bold;"
    >
      Styled 
      <a
        data-testid="styled-link"
        href="/bob"
      >
        @bob
      </a>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`Linkify parses nested HTML and linkifies text nodes 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        Hello 
        <b>
          <a
            data-testid="styled-link"
            href="/alice"
          >
            @alice
          </a>
        </b>
         and 
        <i>
          <a
            data-testid="styled-link"
            href="/explore?q=%23react"
          >
            #react
          </a>
        </i>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      Hello 
      <b>
        <a
          data-testid="styled-link"
          href="/alice"
        >
          @alice
        </a>
      </b>
       and 
      <i>
        <a
          data-testid="styled-link"
          href="/explore?q=%23react"
        >
          #react
        </a>
      </i>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`Linkify renders multiple links and plain text 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      Hi 
      <a
        data-testid="styled-link"
        href="/bob"
      >
        @bob
      </a>
      , check 
      <a
        data-testid="styled-link"
        href="/explore?q=%23js"
      >
        #js
      </a>
       and 
      <a
        data-testid="styled-link"
        href="/carol"
      >
        @carol
      </a>
      !
    </div>
  </body>,
  "container": <div>
    Hi 
    <a
      data-testid="styled-link"
      href="/bob"
    >
      @bob
    </a>
    , check 
    <a
      data-testid="styled-link"
      href="/explore?q=%23js"
    >
      #js
    </a>
     and 
    <a
      data-testid="styled-link"
      href="/carol"
    >
      @carol
    </a>
    !
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`Linkify renders plain text without links 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      Hello world!
    </div>
  </body>,
  "container": <div>
    Hello world!
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
